components/script/dom/byteteeunderlyingsource.rs

```diff
@@ -0,0 +1,457 @@
+    #[ignore_malloc_size_of = "Rc"]
+    reader: Rc<RefCell<ReaderType>>,
+    stream: Dom<ReadableStream>,
+    branch_1: MutNullableDom<ReadableStream>,
+    branch_2: MutNullableDom<ReadableStream>,
+    #[ignore_malloc_size_of = "Rc"]
+    read_again_for_branch_1: Rc<Cell<bool>>,
+    #[ignore_malloc_size_of = "Rc"]
+    read_again_for_branch_2: Rc<Cell<bool>>,
+    #[ignore_malloc_size_of = "Rc"]
+    reading: Rc<Cell<bool>>,
+    #[ignore_malloc_size_of = "Rc"]
+    canceled_1: Rc<Cell<bool>>,
+    #[ignore_malloc_size_of = "Rc"]
+    canceled_2: Rc<Cell<bool>>,
+    #[ignore_malloc_size_of = "Rc"]
+    #[allow(clippy::redundant_allocation)]
+    reason_1: Rc<Box<Heap<Value>>>,
+    #[ignore_malloc_size_of = "Rc"]
+    #[allow(clippy::redundant_allocation)]
+    reason_2: Rc<Box<Heap<Value>>>,
+    #[ignore_malloc_size_of = "Rc"]
+    cancel_promise: Rc<Promise>,
+    tee_cancel_algorithm: ByteTeeCancelAlgorithm,
+    byte_tee_pull_algorithm: ByteTeePullAlgorithm,
+}
+
+impl ByteTeeUnderlyingSource {
+    #[allow(clippy::too_many_arguments)]
+    #[allow(clippy::redundant_allocation)]
+    #[cfg_attr(crown, allow(crown::unrooted_must_root))]
+    pub(crate) fn new(
+        reader: Rc<RefCell<ReaderType>>,
+        stream: &ReadableStream,
+        reading: Rc<Cell<bool>>,
+        read_again_for_branch_1: Rc<Cell<bool>>,
+        read_again_for_branch_2: Rc<Cell<bool>>,
+        canceled_1: Rc<Cell<bool>>,
+        canceled_2: Rc<Cell<bool>>,
+        reason_1: Rc<Box<Heap<Value>>>,
+        reason_2: Rc<Box<Heap<Value>>>,
+        cancel_promise: Rc<Promise>,
+        tee_cancel_algorithm: ByteTeeCancelAlgorithm,
+        byte_tee_pull_algorithm: ByteTeePullAlgorithm,
+        can_gc: CanGc,
+    ) -> DomRoot<ByteTeeUnderlyingSource> {
+        reflect_dom_object(
+            Box::new(ByteTeeUnderlyingSource {
+                reflector_: Reflector::new(),
+                reader,
+                stream: Dom::from_ref(stream),
+                branch_1: MutNullableDom::new(None),
+                branch_2: MutNullableDom::new(None),
+                read_again_for_branch_1,
+                read_again_for_branch_2,
+                reading,
+                canceled_1,
+                canceled_2,
+                reason_1,
+                reason_2,
+                cancel_promise,
+                tee_cancel_algorithm,
+                byte_tee_pull_algorithm,
+            }),
+            &*stream.global(),
+            can_gc,
+        )
+    }
+
+    pub(crate) fn set_branch_1(&self, stream: &ReadableStream) {
+        self.branch_1.set(Some(stream));
+    }
+
+    pub(crate) fn set_branch_2(&self, stream: &ReadableStream) {
+        self.branch_2.set(Some(stream));
+    }
+
+    pub(crate) fn forward_reader_error(&self, this_reader: Rc<RefCell<ReaderType>>, can_gc: CanGc) {
+        let this_reader = this_reader.borrow_mut();
+        match &*this_reader {
+            ReaderType::Default(_reader) => {
+                // TODO
```

Comment:
Is the logic not the same for both readers? Seems like it would be worth making `byte_tee_append_native_handler_to_closed_promise` a function taking all the arguments, including the `closed_promised` from the reader.